[InstCombine] Canonicalize more geps with constant gep bases and constant offsets.

davemgreen opened the pull request:
This is another small but hopefully not performance negative step to canonicalizing towards i8 geps. We looks for geps with a constant offset base pointer of the form `gep (gep @glob, C1), x, C2` and expand the gep
instruction, so that the constant can hopefully be combined together (or the x offset can be computed in common).

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -2822,11 +2822,17 @@ Instruction *InstCombinerImpl::visitGetElementPtrInst(GetElementPtrInst &GEP) {
+                                   m_Shl(m_Value(), m_ConstantInt()))))) ||
+       (isa<GEPOperator>(PtrOp) && isa<ConstantExpr>(PtrOp) &&
+        any_of(drop_begin(GEP.indices()),
+               [](Value *V) { return isa<Constant>(V); })))) {
```
I think this condition has reached the point where it wants to live in its own function :)

I'm also wondering whether zero constants should be excluded from this heuristic, as they don't add any actual offset. And then you wouldn't need the drop_begin? Or at least I don't fully get why the drop_begin() is there.

davemgreen replied:
Thanks - I've pulled out the condition so that it can be more structured. And yeah, the intent for the drop_begin was to rule our the first zero offsets so that is now more explicit too.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineInternal.h:
```diff
@@ -238,6 +238,8 @@ class LLVM_LIBRARY_VISIBILITY InstCombinerImpl final
   std::optional<std::pair<Intrinsic::ID, SmallVector<Value *, 3>>>
   convertOrOfShiftsToFunnelShift(Instruction &Or);
 
+  Value *EmitGEPOffset(GEPOperator *GEP, bool RewriteGEP = false);
```
Why did this move?

davemgreen replied:
Oh - This was a relic of a previous version that moved the creation of geps into (a differently named) shouldCanonicalizeGEPToPtrAdd, but it was simpler to just have the condition in that function.

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -2736,6 +2736,35 @@ Value *InstCombiner::getFreelyInvertedImpl(Value *V, bool WillInvertAllUses,
   return nullptr;
 }
 
+// Return true if we should canonicalize the gep to a i8 ptradd.
```
```suggestion
/// Return true if we should canonicalize the gep to an i8 ptradd.
```

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -2736,6 +2736,35 @@ Value *InstCombiner::getFreelyInvertedImpl(Value *V, bool WillInvertAllUses,
+
+  // gep (gep @global, C1), %x, C2 is expanded so the two constants can
+  // possibly be merged together.
+  return isa<GEPOperator>(PtrOp) && isa<ConstantExpr>(PtrOp) &&
```
Hm, should this one be checking for a constant expression, or that the inner GEPOperator hasAllConstantIndices()? Is there a reason why we want to apply this only to GEPs of globals in particular, rather than all constant GEPs?

davemgreen replied:
I had an older version that was canonicalizing more generally (in a different way IIRC). It was causing some relatively small perf issues so I was trying to avoid making this over-broad. I think that this case should hopefully be OK though, to do the same thing if the geps indices are all constant.

nikic commented:
Looking at https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1382/files it seems like we often don't end up folding the offsets. I think this is because of multi-use limitation in canonicalizeGEPOfConstGEPI8. Possibly it could be relaxed (I think we generally don't mind creating extra constant offset geps...)

nikic reviewed (approved):
LGTM

davemgreen committed:
[InstCombine] Canonicalize more geps with constant gep bases and constant offsets.

This is another small but hopefully not performance negative step to
canonicalizing towards i8 geps. We looks for geps with a constant offset base
pointer of the form `gep (gep @glob, C1), x, C2` and expand the gep
instruction, so that the constant can hopefully be combined together (or the
offset can be computed in common).
